File: car_hw/src/communication.c
#include "communication.h"

#include <stddef.h>
#include <stdbool.h>
#include <usbd_cdc_vcp.h>

#include "memory.h"

void clear_buffer()
{
    uint8_t tmp_char;
    while (VCP_get_char(&tmp_char));
}

void send_int(int n)
{
    VCP_put_char(n >> 24);
    VCP_put_char(n >> 16);
    VCP_put_char(n >> 8);
    VCP_put_char(n);
}

void send_buffer(int size, int pointer)
{
    int i = 0;
    char *buffer = (char *) pointer;

    send_int(size);
    for (; i < size; ++i) {
        VCP_put_char(buffer[i]);
    }
}

int receive_int()
{
    int result = 0;
    uint8_t byte = 0;

    while (!VCP_get_char(&byte));
    result += (byte << 24);
    while (!VCP_get_char(&byte));
    result += (byte << 16);
    while (!VCP_get_char(&byte));
    result += (byte << 8);
    while (!VCP_get_char(&byte));
    result += byte;

    return result;
}

void receive_buffer(int size, int pointer)
{
    char* buffer = (char*) pointer;
    int i = 0;

    for (; i < size; ++i) {
        while(!VCP_get_char(&buffer[i]));
    }
}

File: car_fmw/src/include/communication.kt

external fun VCP_init()
external fun clear_buffer()

external fun send_int(i: Int)
external fun send_buffer(size: Int, pointer: Int)

external fun receive_int(): Int
external fun receive_buffer(size: Int, pointer: Int)

fun receiveByteArray(): ByteArray {
    val result = ByteArray(receive_int())
    receive_buffer(result.size, result.data)

    return result
}

fun sendByteArray(arr: ByteArray) {
    send_buffer(arr.size, arr.data)
}